### PR TITLE
Typed Builder implementation Part 6.5: Shadowing and `require` tests

### PR DESCRIPTION
**Pull request by Kukovec**
- [x] Tests added for any new code
- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [ ] Documentation added for any new functionality
- [ ] Entries added to [./unreleased/][unreleased] for any new functionality

[unreleased]: https://github.com/informalsystems/apalache/tree/unstable/.unreleased

Adds shadowing-guards + tests and better testing coverage of all `require`-guards in builder methods.

### CODE REVIEW

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala [hunk 1]
```diff
@@ -26,9 +26,60 @@ object BuilderUtil {
   /** Removes elem from the scope, as the scope only contains types of free variables */
   def markAsBound(elem: TlaEx): TBuilderInternalState[Unit] = State[TBuilderContext, Unit] { mi: TBuilderContext =>
     require(elem.isInstanceOf[NameEx])
-    (mi.copy(nameScope = mi.nameScope - elem.asInstanceOf[NameEx].name), ())
+    (mi.copy(freeNameScope = mi.freeNameScope - elem.asInstanceOf[NameEx].name), ())
   }
 
+  /** Performs shadowing checks for ternary operators which introduce bound variables */
+  def boundVarIntroductionTernary(
+      rawMethod: (TlaEx, TlaEx, TlaEx) => TlaEx // order: variable, set, expression
```
Comment (shonfeder) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala:34]: I don't think I understand the punctuation in the comment, and I'm not sure I understand the comment either. Doe each word relate to one of the `TlaEx` in the type what is an "order" expression?

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala [hunk 2]
```diff
@@ -26,9 +26,60 @@ object BuilderUtil {
   /** Removes elem from the scope, as the scope only contains types of free variables */
   def markAsBound(elem: TlaEx): TBuilderInternalState[Unit] = State[TBuilderContext, Unit] { mi: TBuilderContext =>
     require(elem.isInstanceOf[NameEx])
-    (mi.copy(nameScope = mi.nameScope - elem.asInstanceOf[NameEx].name), ())
+    (mi.copy(freeNameScope = mi.freeNameScope - elem.asInstanceOf[NameEx].name), ())
   }
 
+  /** Performs shadowing checks for ternary operators which introduce bound variables */
```
Comment (shonfeder) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala:32]: The comment makes it sound like the point of this is to preform shadowing checks, but iiuc, this is meant to construct a particular expression, and the comment is highlighting one particular check that happens during that construction? I think this can be clarified. If the point of the method is to perform these checks, seems like that should be reflected in the method name, or else the comment shouldn't indicate that is the main point?

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala [hunk 3]
```diff
@@ -26,9 +26,60 @@ object BuilderUtil {
...
+  def boundVarIntroductionTernary(
+      rawMethod: (TlaEx, TlaEx, TlaEx) => TlaEx // order: variable, set, expression
+    )(variable: TBuilderInstruction,
+      set: TBuilderInstruction,
+      expr: TBuilderInstruction): TBuilderInstruction = for {
+    setEx <- set
+    usedInSet <- allUsed // x may not appear as bound or free in set
+    exprEx <- expr
+    boundAfterExpr <- allBound // x may not appear as bound in expr
```
Comment (shonfeder) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala:41]: I don't understand these comments. What does `x` refer to? What does the comment have to do with this variable binding?

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala [hunk 4]
```diff
@@ -26,9 +26,60 @@ object BuilderUtil {
...
+    varEx <- variable
+    _ <- markAsBound(varEx)
+    // variable is shadowed iff boundAfterVar \subseteq usedInSet \union boundAfrerExpr
+    boundAfterVar <- allBound
+    diff = (boundAfterVar -- usedInSet) -- boundAfterExpr
```
Comment (shonfeder) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala:46]: Given the comment, why not check for subset relation and then test that boolearather than computing the difference and then checking whether that is empty?

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala (hunk 1 above)
Comment (shonfeder) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala:34]: What are "methods" in this context?

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala [hunk 5]
```diff
@@ -26,9 +26,60 @@ object BuilderUtil {
...
+      rawMethod: (TlaEx, TlaEx) => TlaEx // order: variable, expression
+    )(variable: TBuilderInstruction,
+      expr: TBuilderInstruction): TBuilderInstruction = for {
+    exprEx <- expr
+    boundAfterExpr <- allBound // x may not appear as bound in expr
+    varEx <- variable
+    _ <- markAsBound(varEx)
+    // variable is shadowed iff boundAfterVar \subseteq boundAfrerExpr
+    boundAfterVar <- allBound
+    diff = boundAfterVar -- boundAfterExpr
```
Comment (shonfeder) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala:66]: same question about why you don't just subset here to match the comment.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala [hunk 6]
```diff
@@ -26,9 +26,60 @@ object BuilderUtil {
...
+  /** Performs shadowing checks for binary operators which introduce bound variables */
+  def boundVarIntroductionBinary(
+      rawMethod: (TlaEx, TlaEx) => TlaEx // order: variable, expression
+    )(variable: TBuilderInstruction,
+      expr: TBuilderInstruction): TBuilderInstruction = for {
+    exprEx <- expr
+    boundAfterExpr <- allBound // x may not appear as bound in expr
+    varEx <- variable
+    _ <- markAsBound(varEx)
+    // variable is shadowed iff boundAfterVar \subseteq boundAfrerExpr
```
Comment (shonfeder) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala:64]:
```suggestion
    // variable is shadowed iff boundAfterVar \subseteq boundAfterExpr
```

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/package.scala [hunk 7]
```diff
@@ -38,11 +38,19 @@ package object typecomp {
    * TBuilderContext holds all of the information about the internal state of the builder. It can be extended in the
    * future, to have the builder perform additional static analysis, e.g. assignment analysis.
    *
-   * `nameScope` tracks the types of the variables currently considered as free.
+   *   - `freeNameScope` tracks the types of the variables currently considered as free and their types.
```
Comment (shonfeder) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/package.scala:41]: Seems like "types" appears once too many times here.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/subbuilder/BoolBuilder.scala [hunk 8]
```diff
@@ -29,34 +29,18 @@ trait BoolBuilder extends UnsafeBoolBuilder {
 
   /** \A x \in set: p */
   def forall(x: TBuilderInstruction, set: TBuilderInstruction, p: TBuilderInstruction): TBuilderInstruction =
-    for {
```
Comment (shonfeder) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/subbuilder/BoolBuilder.scala:32]: This is a nice cleanup, even without the added utility of checking for shadowing! :)

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/subbuilder/SetBuilder.scala [hunk 9]
```diff
@@ -44,21 +43,35 @@ trait SetBuilder extends UnsafeSetBuilder {
...
       e: TBuilderInstruction,
       varSetPairs: (TBuilderInstruction, TBuilderInstruction)*): TBuilderInstruction = {
-    val args = varSetPairs.flatMap { case (k, v) =>
-      Seq(k, v)
-    }
-    mapMixed(e, args: _*)
+    for {
+      mapExpr <- e
+      boundAfterMapExpr <- allBound // xi may not appear as bound in mapExpr
+      _ = {}
```
Comment (shonfeder) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/subbuilder/SetBuilder.scala:55]: What is this doing?

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/subbuilder/SetBuilder.scala [hunk 10]
```diff
@@ -44,21 +43,35 @@ trait SetBuilder extends UnsafeSetBuilder {
...
   def map(
       e: TBuilderInstruction,
       varSetPairs: (TBuilderInstruction, TBuilderInstruction)*): TBuilderInstruction = {
-    val args = varSetPairs.flatMap { case (k, v) =>
-      Seq(k, v)
-    }
-    mapMixed(e, args: _*)
+    for {
+      mapExpr <- e
+      boundAfterMapExpr <- allBound // xi may not appear as bound in mapExpr
```
Comment (shonfeder) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/subbuilder/SetBuilder.scala:54]: Again, I don't understand what this comment is doing here. Doesn't seem related to this line and there is no xi mentioned anywhere.

**Review by shonfeder — COMMENTED** [review]
A few typos and some naming/comment questions/suggestions, but LGTM.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala (hunk 4 above)
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala:46]: I mean, they're equivalent. 
```
A \subseteq B <=> A \ B \subseteq B \ B <=> A \ B \subseteq {} <=> A \ B = {}
```

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/subbuilder/SetBuilder.scala (hunk 10 above)
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/subbuilder/SetBuilder.scala:54]: `xi` refers to
https://github.com/informalsystems/apalache/blob/3c9285aedfb2acb11294008f67e5406d20faca3a/tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/subbuilder/SetBuilder.scala#L48
This is actually why the code in `BuilderUtil` mentions `x`, since I originally wrote it for `forall` (`\A x \in S: P`) and then moved it to the utility class when I noticed it could be reused.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/subbuilder/SetBuilder.scala (hunk 9 above)
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/subbuilder/SetBuilder.scala:55]: Skillfully avoiding `fmt-check`. It's leftover code that should be rm'd.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala (hunk 1 above)
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala:34]: `_choose`, `_forall`, `_exists`

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala (hunk 2 above)
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala]: I've added a more verbose description.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala (hunk 4 above)
Comment (Kukovec) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala]: I've changed the tests to be the same as the comments though, might as well.

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala (hunk 1 above)
Comment (shonfeder) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala]: What are those? TLA operators? Scala functions?

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala (hunk 2 above)
Comment (shonfeder) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/BuilderUtil.scala]: Thanks!

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/subbuilder/SetBuilder.scala (hunk 10 above)
Comment (shonfeder) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/subbuilder/SetBuilder.scala:54]: Ah that makes sense. So the comment was a bit not of date. Thanks for the explanation and cleanup.
